CONFIGURATION.md
# Configuration

Chihaya's behaviour is customized by setting up a JSON configuration file.
Available keys are as follows:

##### `httpListenAddr`

    type: string
    default: "localhost:6881"

The listen address for the HTTP server. If only a port is specified, the tracker will listen on all interfaces. If left empty, the tracker will not run a HTTP endpoint.

##### `httpRequestTimeout`

    type: duration
    default: "4s"

The duration to allow outstanding requests to survive before forcefully terminating them.

##### `httpReadTimeout`

    type: duration
    default: "4s"

The maximum duration before timing out read of the request.

##### `httpWriteTimeout`

    type: duration
    default: "4s"

The maximum duration before timing out write of the request.

##### `httpListenLimit`

    type: integer
    default: 0

Limits the number of outstanding requests. Set to `0` to disable.

##### `udpListenAddr`

    type: string
    default: "localhost:6881"

Then listen address for the UDP server. If only a port is specified, the tracker will listen on all interfaces. If left empty, the tracker will not run a UDP endpoint.

##### `createOnAnnounce`

    type: bool
    default: true

Whether to register new torrents with the tracker when any client announces (`true`), or to return an error if the torrent doesn't exist (`false`).

##### `purgeInactiveTorrents`

    type: bool
    default: true

If torrents should be forgotten when there are no active peers.

##### `announce`

    type: duration
    default: "30m"

The announce `interval` value sent to clients. This specifies how long clients should wait between regular announces.

##### `minAnnounce`

    type: duration
    default: "30m"

The announce `min_interval` value sent to clients. This theoretically specifies the minimum allowed time between announces, but most clients don't really respect it.

##### `defaultNumWant`

    type: integer
    default: 50

The default maximum number of peers to return if the client has not requested a specific number.

##### `allowIPSpoofing`

    type: bool
    default: true

Whether peers are allowed to set their own IP via the various supported methods or if these are ignored. This must be enabled for dual-stack IP support, since there is no other way to determine both IPs of a peer otherwise.

##### `dualStackedPeers`

    type: bool
    default: true

True if peers may have both an IPv4 and IPv6 address, otherwise only one IP per peer will be used.

##### `realIPHeader`

    type: string
    default: blank

An optional HTTP header indicating the upstream IP, for example `X-Forwarded-For` or `X-Real-IP`. Use this when running the tracker behind a reverse proxy.

##### `respectAF`

    type: bool
    default: false

Whether responses should only include peers of the same address family as the announcing peer, or if peers of any family may be returned (i.e. both IPv4 and IPv6).

##### `clientWhitelistEnabled`

    type: bool
    default: false

Enables the peer ID whitelist.

##### `clientWhitelist`

    type: array of strings
    default: []

List of peer ID prefixes to allow if `client_whitelist_enabled` is set to true.

##### `torrentMapShards`

    type: integer
    default: 1

Number of internal torrent maps to use. Leave this at 1 in general, however it can potentially improve performance when there are many unique torrents and few peers per torrent.

##### `reapInterval`

    type: duration
    default: "60s"

Interval at which a search for inactive peers should be performed.

##### `reapRatio`

    type: float64
    default: 1.25

Peers will be rated inactive if they haven't announced for `reapRatio * minAnnounce`.

##### `apiListenAddr`

    type: string
    default: "localhost:6880"

The listen address for the HTTP API. If only a port is specified, the tracker will listen on all interfaces. If left empty, the tracker will not run the HTTP API.

##### `apiRequestTimeout`

    type: duration
    default: "4s"

The duration to allow outstanding requests to survive before forcefully terminating them.

##### `apiReadTimeout`

    type: duration
    default: "4s"

The maximum duration before timing out read of the request.

##### `apiWriteTimeout`

    type: duration
    default: "4s"

The maximum duration before timing out write of the request.

##### `apiListenLimit`

    type: integer
    default: 0

Limits the number of outstanding requests. Set to `0` to disable.

##### `driver`

    type: string
    default: "noop"

Sets the backend driver to load. The included `"noop"` driver provides no functionality.

##### `statsBufferSize`

    type: integer
    default: 0

The size of the event-queues for statistics.

##### `includeMemStats`

    type: bool
    default: true

Whether to include information about memory in the statistics.

##### `verboseMemStats`

    type: bool
    default: false

Whether the information about memory should be verbose.

##### `memStatsInterval`

    type: duration
    default: "5s"

Interval at which to collect statistics about memory.


##### `jwkSetURI`

    type: string
    default: ""

If this string is not empty, then the tracker will attempt to use JWTs to validate infohashes before announces. The format for the JSON at this endpoint can be found at [the RFC for JWKs](https://tools.ietf.org/html/draft-ietf-jose-json-web-key-41#page-10) with the addition of an "issuer" key. Simply stated, this feature requires two fields at this JSON endpoint: "keys" and "issuer". "keys" is a list of JWKs that can be used to validate JWTs and "issuer" should match the "iss" claim in the JWT. The lifetime of a JWK is based upon standard HTTP caching headers and falls back to 5 minutes if no cache headers are provided.


#### `jwkSetUpdateInterval`

    type: duration
    default: "5m"

The interval at which keys are updated from JWKSetURI. Because the fallback lifetime for keys without cache headers is 5 minutes, this value should never be below 5 minutes unless you know your jwkSetURI has caching headers.

#### `jwtAudience`

    type: string
    default: ""

The audience claim that is used to validate JWTs.


config/config.go
// Copyright 2015 The Chihaya Authors. All rights reserved.
// Use of this source code is governed by the BSD 2-Clause license,
// which can be found in the LICENSE file.

// Package config implements the configuration for a BitTorrent tracker
package config

import (
	"encoding/json"
	"io"
	"os"
	"time"
)

// Duration wraps a time.Duration and adds JSON marshalling.
type Duration struct{ time.Duration }

// MarshalJSON transforms a duration into JSON.
func (d *Duration) MarshalJSON() ([]byte, error) {
	return json.Marshal(d.String())
}

// UnmarshalJSON transform JSON into a Duration.
func (d *Duration) UnmarshalJSON(b []byte) error {
	var str string
	err := json.Unmarshal(b, &str)
	d.Duration, err = time.ParseDuration(str)
	return err
}

// SubnetConfig is the configuration used to specify if local peers should be
// given a preference when responding to an announce.
type SubnetConfig struct {
	PreferredSubnet     bool `json:"preferredSubnet,omitempty"`
	PreferredIPv4Subnet int  `json:"preferredIPv4Subnet,omitempty"`
	PreferredIPv6Subnet int  `json:"preferredIPv6Subnet,omitempty"`
}

// NetConfig is the configuration used to tune networking behaviour.
type NetConfig struct {
	AllowIPSpoofing  bool   `json:"allowIPSpoofing"`
	DualStackedPeers bool   `json:"dualStackedPeers"`
	RealIPHeader     string `json:"realIPHeader"`
	RespectAF        bool   `json:"respectAF"`
	SubnetConfig
}

// StatsConfig is the configuration used to record runtime statistics.
type StatsConfig struct {
	BufferSize        int      `json:"statsBufferSize"`
	IncludeMem        bool     `json:"includeMemStats"`
	VerboseMem        bool     `json:"verboseMemStats"`
	MemUpdateInterval Duration `json:"memStatsInterval"`
}

// WhitelistConfig is the configuration used enable and store a whitelist of
// acceptable torrent client peer ID prefixes.
type WhitelistConfig struct {
	ClientWhitelistEnabled bool     `json:"clientWhitelistEnabled"`
	ClientWhitelist        []string `json:"clientWhitelist,omitempty"`
}

// TrackerConfig is the configuration for tracker functionality.
type TrackerConfig struct {
	CreateOnAnnounce      bool     `json:"createOnAnnounce"`
	PurgeInactiveTorrents bool     `json:"purgeInactiveTorrents"`
	Announce              Duration `json:"announce"`
	MinAnnounce           Duration `json:"minAnnounce"`
	ReapInterval          Duration `json:"reapInterval"`
	ReapRatio             float64  `json:"reapRatio"`
	NumWantFallback       int      `json:"defaultNumWant"`
	TorrentMapShards      int      `json:"torrentMapShards"`

	NetConfig
	WhitelistConfig
}

// APIConfig is the configuration for an HTTP JSON API server.
type APIConfig struct {
	ListenAddr     string   `json:"apiListenAddr"`
	RequestTimeout Duration `json:"apiRequestTimeout"`
	ReadTimeout    Duration `json:"apiReadTimeout"`
	WriteTimeout   Duration `json:"apiWriteTimeout"`
	ListenLimit    int      `json:"apiListenLimit"`
}

// HTTPConfig is the configuration for the HTTP protocol.
type HTTPConfig struct {
	ListenAddr     string   `json:"httpListenAddr"`
	RequestTimeout Duration `json:"httpRequestTimeout"`
	ReadTimeout    Duration `json:"httpReadTimeout"`
	WriteTimeout   Duration `json:"httpWriteTimeout"`
	ListenLimit    int      `json:"httpListenLimit"`
}

// UDPConfig is the configuration for the UDP protocol.
type UDPConfig struct {
	ListenAddr     string `json:"udpListenAddr"`
	ReadBufferSize int    `json:"udpReadBufferSize"`
}

// Config is the global configuration for an instance of Chihaya.
type Config struct {
	TrackerConfig
	APIConfig
	HTTPConfig
	UDPConfig
	StatsConfig
}

// DefaultConfig is a configuration that can be used as a fallback value.
var DefaultConfig = Config{
	TrackerConfig: TrackerConfig{
		CreateOnAnnounce:      true,
		PurgeInactiveTorrents: true,
		Announce:              Duration{30 * time.Minute},
		MinAnnounce:           Duration{15 * time.Minute},
		ReapInterval:          Duration{60 * time.Second},
		ReapRatio:             1.25,
		NumWantFallback:       50,
		TorrentMapShards:      1,

		NetConfig: NetConfig{
			AllowIPSpoofing:  true,
			DualStackedPeers: true,
			RespectAF:        false,
		},

		WhitelistConfig: WhitelistConfig{
			ClientWhitelistEnabled: false,
		},
	},

	APIConfig: APIConfig{
		ListenAddr:     "localhost:6880",
		RequestTimeout: Duration{10 * time.Second},
		ReadTimeout:    Duration{10 * time.Second},
		WriteTimeout:   Duration{10 * time.Second},
	},

	HTTPConfig: HTTPConfig{
		ListenAddr:     "localhost:6881",
		RequestTimeout: Duration{10 * time.Second},
		ReadTimeout:    Duration{10 * time.Second},
		WriteTimeout:   Duration{10 * time.Second},
	},

	UDPConfig: UDPConfig{
		ListenAddr: "localhost:6882",
	},

	StatsConfig: StatsConfig{
		BufferSize: 0,
		IncludeMem: true,
		VerboseMem: false,

		MemUpdateInterval: Duration{5 * time.Second},
	},
}

// Open is a shortcut to open a file, read it, and generate a Config.
// It supports relative and absolute paths. Given "", it returns DefaultConfig.
func Open(path string) (*Config, error) {
	if path == "" {
		return &DefaultConfig, nil
	}

	f, err := os.Open(os.ExpandEnv(path))
	if err != nil {
		return nil, err
	}
	defer f.Close()

	conf, err := Decode(f)
	if err != nil {
		return nil, err
	}
	return conf, nil
}

// Decode casts an io.Reader into a JSONDecoder and decodes it into a *Config.
func Decode(r io.Reader) (*Config, error) {
	conf := DefaultConfig
	err := json.NewDecoder(r).Decode(&conf)
	return &conf, err
}
